### ISSUE TITLE
Seeking helps on training BigGAN with TPUs

### ISSUE THREAD

**Issue by WingsleyLui** [pytorch/xla#2603]
## ❓ Questions and Help

Hi, I'm seeking helps on training BigGAN on TPU. Following the author's [implementation](https://github.com/ajbrock/BigGAN-PyTorch) and this [repo](https://github.com/giannisdaras/smyrf/tree/master/examples/tpu_biggan) to extend pytorch BigGAN to TPU v3-8. 

Firstly I found it super slow to train: following the setting of [smyrf](https://github.com/giannisdaras/smyrf/tree/master/examples/tpu_biggan), I trained a imagenet with 128x128 resolution with a batch size of 4*8(I think it is quite moderate) . However it turns out that per iteration step costs 8 sec or so to go, this can be seen at [log1](https://drive.google.com/file/d/124UTJ0K3fzVjMBr_fdE2kITRycntlkDa/view?usp=sharing). [Log1](https://drive.google.com/file/d/124UTJ0K3fzVjMBr_fdE2kITRycntlkDa/view?usp=sharing) also recorded some outputs of the train script and what is yielded by `met.metrics_report()`. It turns out that `aten::random_`, `aten::upsample_nearest2d` and `aten::upsample_nearest2d_backward` indicate operations that routed back to CPU engine according to trouble shooting. Finally I followed the solution [#2258](https://github.com/pytorch/xla/issues/2588) to wrap `nn.functional.interpolate` to a module, but this only works for `torch-xla-1.6` but not for the new `torch-xla-1.7`.

For `aten::random_` I did not find any solutions but it comes up from `torch.randint`, `tensor_x.random_(100)`(***this might be a bug to be fixed***), etc. After upsample being good to go, there is a gain on the speed(2~ iteration/sec), and stuff recorded in [log2](https://drive.google.com/file/d/1ZBAXQV7pNDqz6YxJiRdgCjWe-WbE1Yqj/view?usp=sharing).

But still, I'd like to point out that the **matrix_unit_utilization_percent** is quite low: as logged by the monitor: less than 1% before fixing upsample and 10%~ after. When I run the [resnet50 tut](https://cloud.google.com/tpu/docs/tutorials/resnet-pytorch), it was 23% or so. To me it is still quite low, so what infects that utilization, are there any tips to increase that?

Since origin BigGAN used large-batch size and more cores(128~512cores), which is quite costly, I might need some advice to go beyond. Thanks :)

**Comment by JackCaoG** [comment]
@WingsleyLui I will take a look at `random_` issue tomorrow. I think you can use tomorrow's nightly (`torch-xla-nightly`) and `upsample_nearest2d` with `scale_factor` should just work.

**Comment by WingsleyLui** [comment]
> @WingsleyLui I will take a look at `random_` issue tomorrow. I think you can use tomorrow's nightly (`torch-xla-nightly`) and `upsample_nearest2d` with `scale_factor` should just work.

Sounds great~! Apart from this, any ideas on how to enhance matrix_unit_utilization_percent?

**Comment by JackCaoG** [comment]
@WingsleyLui We did not have `the lowering for `aten::random_` which is the discrete uniform distribution now. You can workaround it with `uniform` like

```
torch.FloatTensor(dims).uniform_(low, high).round()
``` 
or
```
((high - low) * torch.rand(dims)).round() + low
```
for now.

I will see if there is an easy way for us to lower this op.

**Comment by WingsleyLui** [comment]
Thanks. Will give it a shot :)

**Comment by foobar8675** [comment]
@WingsleyLui just curious, did you ever get bigan working on a tpu? (I'm trying to get stylegan working)

**Comment by WingsleyLui** [comment]
hi @foobar8675. [smyrf](https://github.com/giannisdaras/smyrf/tree/master/examples/tpu_biggan) got biggan work on CelebA dataset. I gave it a shot on this implementation before, but as this issue pointed out, there are some lowerings  which slow down the process:

1. The upsample operation switches between cpu and tpu, which is super slow and lead to 0% utility on tpu(idle). (the team claims that they've solved this problem in the latest version).
2. The discrete random sample process also operation switches between cpu and tpu, but this can be ignored. it seems that stylegan is unconditional so I think it does not matter.
3. If you use gradient accumulation to mimic a large batch size, I would lead to 0% utility on tpu(idle), for which I did not figure out.
4. You may keep training GAN when using tpu since eval on FID/IS would change the graph, thus leading to 0% utility(idle) on tpu.
5. You need to apply for additional VMs to scale up if you need multi-tpus, this is much more expensive than TF.

So based on my experience, only trials on simple training on GANs could be done smoothly on PyTorch-xla. I did not finish the whole training process cuz pytorch-xla is not very mature currently.

**Comment by foobar8675** [comment]
@WingsleyLui got it. thanks for the detailed notes on your finding. i did notice some of the same as well, which has been ... challenging...  I'm going to keep trying to work around stylegan for a little longer, but there's a good chance I'll end up with the same conclusion as yourself. :(
